### ISSUE TITLE
GA BLOG: SEC-381 Trusted certificate enhancements

### ISSUE THREAD

**Issue by acdemyers** [OpenLiberty/open-liberty#9918]
a. Which Liberty feature does it relate to?
   i. Human-readable name (eg WebSockets feature)
      **Transport Security 1.0**
   ii. Feature name (eg websockets-1.0)
   **transportSecurity-1.0**
   iii. Is this a new feature or is it an update to an existing feature for this release?

    Update to the existing transportSecurity-1.0 feature.
b. Who is the target audience? eg application developers, devops pipeline devs, appserver administrators, etc

   Application developers and appserver administrators
c. Write a sentence or two that introduces it to someone new to the general technology/concept.
 
The Liberty server is offering a new options to help user establish trust for SSL connections.

d. What is new/changed in this release and why should they get excited about this new thing (2-3 sentences)? What was the problem before and how this new feature make their life easier? (Why should they care?)

With this feature users has options to help them establish trust for SSL communications.   Users are provided an easy way to use the JDK's default truststore, likely the cacerts, for trust.  Users are also provide with a way to pass the certificate needed to establish trust to a truststore through an environment variable.
 
e. Explain how to make it work. Include screenshots and/or code snippets to illustrate how to use it, if relevant. Provide server.xml snippet for how to enable the feature.

1.  Establishing trust using the JDK's default truststore.

Liberty can be configured to use the JDK's default truststore to establish trust.  By default the JDK default truststore is the cacerts file.  But the default truststore will be what is set by the javax.net.ssl.truststore system property if set or may be the jssecacerts file if users have one configured.    For Liberty to use what is configured as the JDK default truststore the "useDefaultCerts" attribute needs to be set to true on the "ssl" element, it is set to false by default.   Example configuration:

	<ssl id="defaultSSLConfig" keyStoreRef="defaultKeyStore" trustStoreRef="defaultTrustStore" trustDefaultCerts="true" />
	<keyStore id="defaultKeyStore" location="key.p12" type="PKCS12" password=<your password>  />
         <keyStore id="defaultTrustStore" location="trust.p12" type="PKCS12" password=<your password> />

With "trustDefaultCerts" set to true,  the Liberty server will try to establish trust with the configured truststore in this case "defaultTrustStore" first.  If trust is not establish with the configured truststore then it will try to use the JDK's default truststore to establish trust.  

2.  Providing a certificate through an environment variable to establish trust

Liberty will read a certificate from an environment variable and add it to a keystore or truststore so it can be used for trust.  The certificate will be added to the runtime copy of the keystore or truststore and will not be stored to the file system.  If the keystore configuration include the "readOnly" attribute set to true then the certificate will not be included. 

The environment variable key must be in the format  "cert_" + keystore id.   For a keystore configuration that looks like this:

       <keyStore id="myKeyStore" location="myKey.p12" type="PKCS12" password=<your password> />

The key of the environment variable should be "cert_myKeyStore",  it is case sensitive.

The value of the environment variable can either be a certificate in the base 64-bit format or the path to a file containing a base 64-bit encode certificate or DER encoded certificate.   If using the base 64-bit encode certificate directly on the environment variable it must contain the "-----BEGIN CERTIFICATE-----" and "-----END CERTIFICATE-----" tags.   eg

**cert_myKeyStore="-----BEGIN CERTIFICATE-----
....
-----END CERTIFICATE-----"**

For a file the environment variable will look similar to:

**cert_myKeyStore=/Users/me/servercert.crt**

Any value not starting with the "-----BEGIN CERTIFICATE-----" tag will be treated like a file.

f. Provide any public URLs where the reader can find out more about this specific update and/or about the wider technology.
g. Attach any images to this work item as .png files.

**Comment by lmsurpre** [comment]
Not sure if its better to leave this here or open a new issue, but I think there's an unfortunate mistake in the blog post which seems to be the only documentation of the cert enhancements.

First it says to set the `useDefaultCerts` attribute: 
> For Liberty to use what is configured as the JDK default truststore the "useDefaultCerts" attribute needs to be set to true on the "ssl" element, it is set to false by default. Example configuration:

Then it uses `trustDefaultCerts` as the attribute name in the sample config (and the paragraph that follows):
```
<ssl id="defaultSSLConfig" keyStoreRef="defaultKeyStore" trustStoreRef="defaultTrustStore" trustDefaultCerts="true" />
<keyStore id="defaultKeyStore" location="key.p12" type="PKCS12" password=<your password>  />
<keyStore id="defaultTrustStore" location="trust.p12" type
```

I think that `trustDefaultCerts` is the only valid attribute name (and that `useDefaultCerts` is wrong), but please correct me if I'm wrong.

**Comment by lauracowen** [comment]
@acdemyers ^^^

**Comment by acdemyers** [comment]
@lauracowen   The correct name of the attribute is trustDefaultCerts

**Comment by brutif** [comment]
@lauracowen looks like this never made the blog, but I think the part about providing certificates through an environment variable or a file could be an example at https://openliberty.io/docs/ref/feature/#transportSecurity-1.0.html.  It's a very useful feature for container environments that hardly anyone knows about.  Could you let @acdemyers and I know how to submit that update?  Then we could probably close this issue.

**Comment by lauracowen** [comment]
This blog post? https://openliberty.io/blog/2019/12/06/microprofile-32-health-metrics-190012.html#ssl

Oh you mean the comment in Feb. No, we didn't see that - we don't look back at the issues once the post draft is done (this issue should've been closed, sorry). Good spot.

Can you create a PR following this process here, please: https://github.com/openliberty/blogs#Authors-and-Editors-updating-a-published-post Thx.

**Comment by jakub-pomykala** [comment]
Closing. Blog post was update with the comment. (Verified by checking it myself)
